developer-docs-site/docs/concepts/upgrading-code.md

```diff
@@ -0,0 +1,62 @@
+---
+title: "Upgrading Code"
+slug: "upgrading"
+---
+import ThemedImage from '@theme/ThemedImage';
+import useBaseUrl from '@docusaurus/useBaseUrl';
+
+# Upgrading Code
+
+The Aptos chain supports _code upgrade_, which means already deployed Move
+code can be replaced with newer versions. Code upgrade enables
+code owners to evolve their contracts or frameworks under a stable, well-known
+account address, which can than be referenced by other applications, inside or
+outside the chain.
+
+## How it Works
+
+Code upgrade on the Aptos chain happens on [Move package]
+(https://move-language.github.io/move/packages.html) level. A package
+in Aptos Move specifies an _upgrade policy_ in the manifest:
+
+```toml
+[package]
+name = "MyApp"
+version = "0.0.1"
+upgrade_policy = "compatible"
+```
+
+The given policy means that any upgrades to this package must be _downwards
+compatible_ both regards storage and apis, conditions which can be checked
+mechanical from the Move bytecode:
+
+- For storage, all old struct declarations must be in exactly the same way
+  in the new code. This ensures the existing state of storage is correctly
+  interpreted by the new code. Indeed, new struct declarations can be added.
+- Similar for apis: all public functions must have the same signature as  
+  before, but new ones can be added.
+
+Aptos checks compatibility at the time a [Move package](https://move-language.github.io/move/packages.html) is published via a
+dedicated Aptos framework transaction. This transaction aborts if
+compatibility is not satisfied as specified.
+
+### Different Upgrade Policies
+
+Besides the upgrade policy `compatible`, a package owner can also choose
+`immutable` (disallowing any future upgrade), or `arbitrary` (performing no
+checks at all). The policy of a given package can start as `arbitrary`,  
+then move to `compatible`, and then move to `immutable` -- but it can never
```

Comment:
I thought we were planning on removing arbitrary or not making it available on mainnet?

Reply:
We were planning on discussing this. I see no reason to remove it. There is nothing it can destroy on chain. Let users decide what they want to do. Let communities experiment with different notions of compatibility. (Yes, the current one is not the only one, and tools can compute compatibility based on the (new) ABI.)

Reply:
Conservatively module upgrades are rather tricky. Even without the arbitrary nature, many uncomfortable scenarios could happen. I do worry about us not knowing enough about the community needs and giving them too much flexibility or responsibility. I think another goal was to make this only possible for applications and not for libraries and I don't now if that has yet to be implemented. I'd at least prefer we start by not supporting it and evaluating over time if we want to go toward that direction. Preferably let's get more experience what it means to support module upgrades and evaluate.